src/python/pants/option/optionable.py

```diff
@@ -41,9 +43,19 @@ def signature(cls):
 
     TODO: This indirection avoids a cycle between this module and the `rules` module.
     """
-    snake_scope = cls.options_scope.replace('-', '_')
     partial_construct_optionable = functools.partial(_construct_optionable, cls)
+
+    # NB: We must populate several dunder methods on the partial function because partial functions
+    # do not have these defined by default and the engine uses these values to visualize functions
+    # in error messages and the rule graph.
+    snake_scope = cls.options_scope.replace('-', '_')
     partial_construct_optionable.__name__ = f'construct_scope_{snake_scope}'
+    partial_construct_optionable.__module__ = cls.__module__
+    # It's a code smell to use unittest.mock in production, but this is a lightweight mechanism
+    # to get the engine to correctly display the class definition line.
+    _, class_definition_lineno = inspect.getsourcelines(cls)
+    partial_construct_optionable.__code__ = Mock(co_firstlineno=class_definition_lineno)
```

Comment:
This is wonky, but the best I could come up with after trying these different approaches:

1) Create our own mock object. Far more verbose and probably less performant than simply using `unittest.mock`.
2) Change the Rust implementation to only render the line number if available. Wasn't sure how to do this and I don't think it's ideal because the line number is always useful for debugging.
3) To avoid using `inspect`, use `cls.__init__.__code__.co_firstlineno`. This doesn't work because Subsystems and GoalSubsystems never are supposed to override `__init__` so the lineno referred to the superclasses.

The end result is very neat, though!

```
pants.rules.core.list_targets:14:construct_scope_list
pants.rules.core.binary:18:construct_scope_binary
pants.backend.python.subsystems.subprocess_environment:12:construct_scope_subprocess_environment
```